# Introduction

This Data Science Project analyzes data from a dating application called OKCupid. As a rise in dating apps continues, the amount fo data collected has also rised. This data can be used to gain insights and find hidden traits that would have been otherwise impossible to find without using Data Science. 

The goals of this project are to scope, prep, analyze and build a machine learning model from it 

**Data Sources**

`profiles.csv` provided by Codecademy.com (which I have not added to the git repository due to file size restrictions)

### Project Goals

The primary goal of this project to apply skills learned through the Code Academy Data Science career path and apply machine learning techniques to a dataset. The question in focus is to predict user's astrological sign from the given variables from their profiles. 

### Data

The data is provided by Codecademy in `profiles.csv`. The rows contain individual users while the columns represent the features from thier prodiles. 

### Analysis

This project will use descriptive statistics and data visualization to find key features and relationships between variables. Since we have to predict user's astrological signs, a classification algorithm will be used from the supervised ML models.

### Evaluation

The project will conclude with evaluating the models score through confusion matrix and metrics such as accuracy, precision, recall, and f1 scores.



### Importing required libraries

In [17]:
import pandas as pd
from matplotlib import pyplot as plt
import seaborn as sns
import numpy as np 
pd.set_option('display.max_columns', None)

### Load the dataset


In [18]:
data = pd.read_csv('profiles.csv')
data.head()

age       body_type               diet    drinks      drugs  \
0   22  a little extra  strictly anything  socially      never   
1   35         average       mostly other     often  sometimes   
2   38            thin           anything  socially        NaN   
3   23            thin         vegetarian  socially        NaN   
4   29        athletic                NaN  socially      never   

                           education  \
0      working on college/university   
1              working on space camp   
2     graduated from masters program   
3      working on college/university   
4  graduated from college/university   

                                              essay0  \
0  about me:<br />\n<br />\ni would love to think...   
1  i am a chef: this is what that means.<br />\n1...   
2  i'm not ashamed of much, but writing public te...   
3          i work in a library and go to school. . .   
4  hey how's it going? currently vague on the pro...   

                                              essay1  \
0  currently working as an international agent fo...   
1  dedicating everyday to being an unbelievable b...   
2  i make nerdy software for musicians, artists, ...   
3          reading things written by old dead people   
4                         work work work work + play   

                                              essay2  \
0  making people laugh.<br />\nranting about a go...   
1  being silly. having ridiculous amonts of fun w...   
2  improvising in different contexts. alternating...   
3  playing synthesizers and organizing books acco...   
4  creating imagery to look at:<br />\nhttp://bag...   

                                              essay3  \
0  the way i look. i am a six foot half asian, ha...   
1                                                NaN   
2  my large jaw and large glasses are the physica...   
3                  socially awkward but i do my best   
4            i smile a lot and my inquisitive nature   

                                              essay4  \
0  books:<br />\nabsurdistan, the republic, of mi...   
1  i am die hard christopher moore fan. i don't r...   
2  okay this is where the cultural matrix gets so...   
3  bataille, celine, beckett. . .<br />\nlynch, j...   
4  music: bands, rappers, musicians<br />\nat the...   

                                              essay5  \
0  food.<br />\nwater.<br />\ncell phone.<br />\n...   
1  delicious porkness in all of its glories.<br /...   
2  movement<br />\nconversation<br />\ncreation<b...   
3                                                NaN   
4                                                NaN   

                        essay6  \
0  duality and humorous things   
1                          NaN   
2                          NaN   
3   cats and german philosophy   
4                          NaN   

                                              essay7  \
0  trying to find someone to hang out with. i am ...   
1                                                NaN   
2  viewing. listening. dancing. talking. drinking...   
3                                                NaN   
4                                                NaN   

                                              essay8  \
0  i am new to california and looking for someone...   
1  i am very open and will share just about anyth...   
2  when i was five years old, i was known as "the...   
3                                                NaN   
4                                                NaN   

                                              essay9            ethnicity  \
0  you want to be swept off your feet!<br />\nyou...         asian, white   
1                                                NaN                white   
2  you are bright, open, intense, silly, ironic, ...                  NaN   
3                              you feel so inclined.                white   
4                                                NaN  asian, black, other   

 

### Understanding the data

We can extract some key information from the data by finding the columns present, the # of rows and if column names need to be renamed for easier access

In [19]:
print(data.columns)
print(data.shape)

Index(['age', 'body_type', 'diet', 'drinks', 'drugs', 'education', 'essay0',
       'essay1', 'essay2', 'essay3', 'essay4', 'essay5', 'essay6', 'essay7',
       'essay8', 'essay9', 'ethnicity', 'height', 'income', 'job',
       'last_online', 'location', 'offspring', 'orientation', 'pets',
       'religion', 'sex', 'sign', 'smokes', 'speaks', 'status'],
      dtype='object')
(59946, 31)


From the information above we can see that there is plenty of data available to use in a machine learning model for training.
There are 31 columns and 59946 rows

As we are interested in the astrological signs of the users, lets see that column and make any necessary cleaning

In [20]:
print(data['sign'])
data['sign_cleaned'] = data['sign'].str.split().str.get(0)
print(data['sign_cleaned'])

0                                          gemini
1                                          cancer
2              pisces but it doesn&rsquo;t matter
3                                          pisces
4                                        aquarius
                           ...                   
59941    cancer and it&rsquo;s fun to think about
59942             leo but it doesn&rsquo;t matter
59943     sagittarius but it doesn&rsquo;t matter
59944       leo and it&rsquo;s fun to think about
59945    gemini and it&rsquo;s fun to think about
Name: sign, Length: 59946, dtype: object
0             gemini
1             cancer
2             pisces
3             pisces
4           aquarius
            ...     
59941         cancer
59942            leo
59943    sagittarius
59944            leo
59945         gemini
Name: sign_cleaned, Length: 59946, dtype: object


In [21]:
print(data.sign_cleaned.nunique())
print(data.sign_cleaned.unique())

12
['gemini' 'cancer' 'pisces' 'aquarius' 'taurus' 'virgo' 'sagittarius'
 'leo' nan 'aries' 'libra' 'scorpio' 'capricorn']


Now that we have cleaned the data for our use, visualizing the data would be essential to see what features correlate. That will allow us to remove any unwanted features that would produce noise